### DISCUSSION TITLE
Rerender problem with inputs (HELP NEEDED)

### DISCUSSION THREAD

**Discussion by 7iomka** [mantinedev/mantine#3554] [Q&A] [answered]
Hi.
I use mantine inputs for my project, and I made component wrappers for each type of input to meet the requirements of my design.
There is an example for TextInput.

https://codesandbox.io/s/aged-wood-t2rm44?file=/src/components/text-input/text-input.tsx

I have one big problem now - when parent component changes, my TextInput also changes, even if value / defaultValue is not changed, and even if it wrapped with memo. Open the console and see logs inside TextInput when you click on button that increments counter.

Maybe I made some mistake in the construction of the wrapper component?
I would be very grateful for your advice.

For me this problem is critical at the moment, because inputs are used in wizards where the parent component changes, which forces the input to be re-rendered and triggers unnecessary validation (I used my own form lib).

Many thanks!

**Comment by rtivital** [comment, marked as answer]
That is how React works – if any state in the parent component is changed and children depend on any part of this state, then they re-render, it does not matter whether the state was actually changed. It is very time-consuming to debug this, you can use why-did-you-update or similar library to find out the cause of re-rendering.

**Comment by 7iomka** [comment]
@rtivital in case of this input, it value not depends on parent data.
I installed `why-did-you-update` lib
https://codesandbox.io/s/aged-wood-t2rm44?file=/src/components/text-input/text-input.tsx

See logs in browser.
How I see the we have multiple logs from lib with single message - the theme object passed for MantineProdiver is not equal between rerenders, but it actually equal.
It is possible somehow to determine the problem?

**Reply by rtivital** [reply]
I am not sure about that. There is no memoization in theme merging, this might be the case for some re-renders. You can find theme related source code here – https://github.com/mantinedev/mantine/blob/v6/src/mantine-styles/src/theme/MantineProvider.tsx#L81:L84

**Comment by 7iomka** [comment]
@rtivital 
How I see after research, this issue is not relating with theme merging, rerenders happens even you have just provider.
If you don't use MantineProvider - you have not this rerenders.

Demo: https://codesandbox.io/s/nifty-albattani-dieprm?file=/src/App.tsx

**Reply by rtivital** [reply]
I suppose that it is related to theme merging because this logic is in MantineProvider. There are only 3 parts that can cause the issue:
- ThemeProvider from emotion (I doubt about that)
- Provider context (possibly)
- Theme merging (possibly)

**Reply by 7iomka** [reply]
> I suppose that it is related to theme merging because this logic is in MantineProvider. There are only 3 parts that can cause the issue:
> 
> * ThemeProvider from emotion (I doubt about that)
> * Provider context (possibly)
> * Theme merging (possibly)

This is sad news, because I have unexpected rerenders for most of components because of that.
And I don't know how to deal with....

**Reply by rtivital** [reply]
You can identify the issue by exploring source code, then you can send a PR that will fix it
